docs/RingBuffer.md

```diff
@@ -4,6 +4,537 @@ ebpf-for-windows exposes the [libbpf.h](/include/bpf/libbpf.h) interface for use
+2. Call `ebpf_ring_buffer_set_wait_handle` to set the wait handle.
+3. Directly read records from the producer pages (and update consumer offset as we read).
+    - Use acquire and release semantics as described below for accessing record headers and ring offsets.
+4. Call `WaitForSingleObject`/`WaitForMultipleObject` as needed to wait for new data to be available.
+
+### Differences from linux API
+
+#### Poll and Consume
+
+On linux `ring_buffer__poll()` and `ring_buffer__consume()` are used to invoke the callback.
+`poll()` waits for available data (or until timeout), then consume all available records.
+`consume()` consumes all available records (without waiting).
+
+Windows will initially only support `ring_buffer__poll()`, which can be called with a timeout of zero
+to get the same behaviour as `ring_buffer__consume()`.
+
+#### Asynchronous callbacks
+
+On Linux ring buffers currently support only synchronous callbacks (using poll/consume).
+In contrast, Windows eBPF currently supports only asynchronous ring buffer callbacks,
+where the callback is automatically invoked when data is available.
+
+This proposal adds support for synchronous consumers by setting the `RINGBUF_FLAG_NO_AUTO_CALLBACK` flag.
+With the flag set, callbacks will not automatically be called.
+To invoke the callback and `ring_buffer__poll()`
+should be called to poll for available data and invoke the callback.
+On Windows a timeout of zero can be passed to `ring_buffer__poll()` to get the same behaviour as `ring_buffer__consume()` (consume available records without waiting).
+
+When #4142 is resolved the default behaviour will be changed from asynchronous (automatic) to synchronous callbacks,
+so `RINGBUF_FLAG_AUTO_CALLBACK` should always be specified for asynchronous callbacks for forward-compatibility.
+
+#### Memory mapped consumers
+
+As an alternative to callbacks, Linux ring buffer consumers can directly access the
+ring buffer data by calling `mmap()` on a ring_buffer map fd to map the data into user space.
+`ring_buffer__epoll_fd()` is used on Linux to get an fd to use with epoll to wait for data.
+
+Windows doesn't have directly compatible APIs to Linux mmap and epoll, so instead we will perfom the mapping
+in the eBPF core and use a KEVENT to signal for new data.
+
+For direct memory mapped consumers on Windows, use `ebpf_ring_buffer_get_buffer` to get pointers to the producer and consumer
+pages mapped into user space, and `ebpf_ring_buffer_get_wait_handle()` to get the SynchronizationEvent (auto-reset) KEVENT
+to use with `WaitForSingleObject`/`WaitForMultipleObject`.
+
+Similar to the linux memory layout, the first page of the producer and consumer memory is the "producer page" and "consumer page",
+which contain the 64 bit producer and consumer offsets as the first 8 bytes.
+Only the producer may update the producer offset, and only the consumer may update the consumer offset.
```

Comment:
Clarify: is this statement enforced somehow or is it merely a convention?
"may" sounds like it's enforced.  If merely a convention, change both uses of "may" to "should"